Add support for Kotlin `@Throws` in metadata

ZacSweers opened the pull request:
This adds support for detecting `@Throws` in metadata. Like some other jvm-specific things (certain modifiers, etc), these have to be inferred from the jvm code as the annotations are not present

Egorand commented on kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt:
```diff
@@ -129,6 +130,21 @@ interface ElementHandler {
+   * @param classJvmName The JVM name of the class (example: `"org/foo/bar/Baz$Nested"`).
+   * @param methodSignature The method with annotations to look up.
+   * @param isConstructor Indicates if [methodSignature] is a constructor.
+   * @return the set of found thrown types, or null.
```
nit: Any upside in returning null vs an empty set?

ZacSweers replied:
null allows us to lean on kotlin nullability rather than checking if the returned annotations are empty

JakeWharton replied:
Collections should almost never be nullable unless there's some semantic meaning between non-null vs. null and empty vs. non-empty. For things like an allow/disallow list, null might mean allow all whereas empty list means allow nothing.

Since you don't need the ternary state, empty list makes more sense here. You already always have the problem of dealing with empty lists in collections. Adding null just complicates things.

ZacSweers replied:
fair enough, will change 👍

Egorand commented on kotlinpoet-metadata-specs-tests/src/test/kotlin/com/squareup/kotlinpoet/metadata/specs/test/KotlinPoetMetadataSpecsTest.kt:
```diff
@@ -1192,6 +1192,47 @@ class KotlinPoetMetadataSpecsTest(
+
+    //language=kotlin
+    assertThat(typeSpec.trimmedToString()).isEqualTo("""
+      class Throwing {
```
Constructor exception didn't get recognized?

ZacSweers replied:
oooo good catch! Will look

Egorand commented on kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/KotlinPoetMetadataSpecs.kt:
```diff
@@ -816,6 +826,19 @@ private inline fun <E> setOf(body: MutableSet<E>.() -> Unit): Set<E> {
+  return AnnotationSpec.builder(Throws::class)
+      .addMember(
+          "exceptionClasses = %L",
+          exceptions.map { CodeBlock.of("%T::class", it) }
```
nit: These won't get new lines automatically if the array is too long, this would require some manual formatting. Not important for this PR.

ZacSweers commented:
Fixed missing constructor annotations and changed to use a non-nullable return type.

I also found a bug in `ParameterSpec` where primary constructors with no keyword emission would still emit parentheses even with no parameters. Technically legal in kotlin, but redundant

Egorand commented on kotlinpoet/src/main/java/com/squareup/kotlinpoet/ParameterSpec.kt:
```diff
@@ -183,9 +183,13 @@ class ParameterSpec private constructor(
 internal fun List<ParameterSpec>.emit(
   codeWriter: CodeWriter,
   forceNewLines: Boolean = false,
+  forceParensOnEmpty: Boolean = true,
```
I'd really prefer these changes to be in a separate PR, looks they're not strictly related to this one. Can you please pull them out and rebase this PR on top? Should be an easy merge.

ZacSweers replied:
https://github.com/square/kotlinpoet/pull/774

Egorand commented on kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt:
```diff
@@ -473,7 +473,7 @@ class TypeSpec private constructor(
     fun primaryConstructor(primaryConstructor: FunSpec?) = apply {
       check(kind == Kind.CLASS) {
-        "$kind can't have initializer blocks"
+        "$kind can't have a primary constructor"
```
Also feels like a separate PR, even if a very small one.

ZacSweers replied:
https://github.com/square/kotlinpoet/pull/773

Egorand commented:
Looks good! Wanna rebase against master before I merge?

ZacSweers commented:
Rebased!
